### PR TITLE
Store Pending Deletions Fix

### PR DESCRIPTION
**Pull request by henningandersen**
FilterDirectory.getPendingDeletions does not delegate, fixed
temporarily by overriding in StoreDirectory.

This in turn caused duplicate file name use after a trimUnsafeCommits
had been done, since a new IndexWriter would not consider the pending
deletes in IndexFileDeleter. This should only happen on windows (AFAIK).

Reenabled doing index updates for all tests using
IndexShardTests.indexOnReplicaWithGaps (which could fail due to above
when using mocked WindowsFS).

Follow-up to #40249

A follow-up should fix this in FilterDirectory.

### CODE REVIEW

**Review by ywelsch — APPROVED** [review]
Great catch! LGTM

**Comment by dnhatn** [comment]
@henningandersen I also dug into a recent test failure of testRestoreLocalHistoryFromTranslogOnPromotion (https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-unix-compatibility/os=centos-6/312/console). I think that test failure has the same cause as the one you are fixing here. I proposed https://issues.apache.org/jira/browse/LUCENE-8734 to record nextWriteDVGen.

**Comment by henningandersen** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample

File: server/src/main/java/org/elasticsearch/index/store/Store.java [hunk 1]
```diff
@@ -760,6 +760,13 @@ public IndexInput openInput(String name, IOContext context) throws IOException {
         public String toString() {
             return "store(" + in.toString() + ")";
         }
+
+        @Override
+        public Set<String> getPendingDeletions() throws IOException {
+            // FilterDirectory.getPendingDeletions does not delegate, working around it here.
+            // to be removed once fixed in FilterDirectory.
+            return unwrap(this).getPendingDeletions();
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/index/store/Store.java]: I think this fix is trappy, but I am ok with it for now. I would like us to delegate all our subclasses of `FilterDirectory` as well. We have a few more. I am torn on the unwrap to be honest.

Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/index/store/Store.java]: The problem with that route is that most of the subclasses of this are in lucene, including the `MockDirectoryWrapper` created by `LuceneTestCase` and for instance `HardlinkCopyDirectoryWrapper` used by `StoreRecovery`.
I agree that using `unwrap` is questionable, but in the given context this should be safe here until fixed in lucene.

**Comment by henningandersen** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample

**Comment by henningandersen** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample

**Review by dnhatn — APPROVED** [review]
Great find!

File: server/src/main/java/org/elasticsearch/index/store/Store.java (hunk 1 above)
Comment (s1monw) [reply server/src/main/java/org/elasticsearch/index/store/Store.java]: I agree, we should still pro-actively implement the right fix in all subclasses and remove the unwrap once lucene is upgraded

Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/index/store/Store.java]: Thanks @s1monw  for clarifying. Except for two trivial test-only subclasses, I have added the override to all subclasses in 85f2066ec34b1e400d098d3ae0c742ee73288dfe.

Please have another look at your convenience.

**Review by s1monw — APPROVED** [review]

**Pull request merged**